src/buf_writer.c

```diff
@@ -5,8 +5,43 @@
 #include "jemalloc/internal/buf_writer.h"
 #include "jemalloc/internal/malloc_io.h"
 
+static void *
+buf_writer_allocate_internal_buf(size_t buf_len) {
+	if (buf_len > SC_LARGE_MAXCLASS) {
+		return NULL;
+	}
+	return iallocztm(TSDN_NULL, buf_len, sz_size2index(buf_len), false,
+	    NULL, true, arena_get(TSDN_NULL, 0, true), true);
+}
+
+static void
+buf_writer_free_internal_buf(void *buf) {
+	idalloctm(TSDN_NULL, buf, NULL, NULL, true, true);
+}
+
+bool
+buf_writer_init(buf_writer_t *buf_writer, write_cb_t *write_cb, void *cbopaque,
+    char *buf, size_t buf_len) {
+	buf_writer->write_cb = write_cb;
+	buf_writer->cbopaque = cbopaque;
+	assert(buf_len >= 2);
+	if (buf != NULL) {
```

Comment:
It looks the buffer is never passed in. Do you have any use case in mind where the buffer is external?

Reply:
Yes, when a static buffer is preferred. An example is profiling, where we statically allocate the buffer, and this buffered writer should be able to accept it. (#1527 is trying to migrate profiling to use the buffered writer.) Currently the stats printing and profiling last-N dumping request a dynamic buffer, but we may change it later.

In my unit test I'm testing three cases: a static buffer, a dynamic buffer, and OOM.